### PR TITLE
Notification request fix

### PR DESCRIPTION
**Pull request by JoannesJ**

### CODE REVIEW

**Comment by xDashh** [comment]
That's a big PR, need @KenanY or @slugalisk to look over it.
Only thing that makes me curious is that afaik you are supposed to check the origin of window messages (like you are using for communication of frames) and i don't see that happening atm.

**Comment by JoannesJ** [comment]
You can pass a `targetOrigin` to `postMessage`, but all the popup does is signal that it's done.
There is no security concern there.

File: assets/chat/js/menus.js [hunk 1]
```diff
@@ -168,31 +169,41 @@ class ChatSettingsMenu extends ChatMenu {
...
-                case 'granted':
-                    resolve(Notification.permission);
-                    break;
-                case 'denied':
-                default:
-                    reject(Notification.permission);
-                    break;
+            // Detecting for Blink webengine and whether we are not the top frame
+            // See issue github.com/MemeLabs/chat-gui/issues/12
+            const isBlink = window.Intl && Intl.v8BreakIterator;
```
Comment (KenanY) [inline assets/chat/js/menus.js]: I think `Intl` will eventually be supported by all browsers, and `Intl.v8BreakIterator` is unofficial and slated for removal. We will need to find another way to detect Blink.

File: assets/chat/js/menus.js [hunk 2]
```diff
@@ -168,31 +169,41 @@ class ChatSettingsMenu extends ChatMenu {
...
+                );
+
+                const onMessage = event => {
+                    if( event.data.name === "notification-request-done" ){
+                        window.removeEventListener('message', onMessage);
+                        resolve(Notification.permission);
+                    }
+                };
+
+                window.addEventListener( 'message', onMessage );
```
Comment (KenanY) [inline assets/chat/js/menus.js]: Other function calls in this file don't add spaces before and after the list of parameters.

File: assets/chat/js/menus.js [hunk 3]
```diff
@@ -168,31 +169,41 @@ class ChatSettingsMenu extends ChatMenu {
...
+            } else {
+                Notification.requestPermission( resolve );
```
Comment (KenanY) [inline assets/chat/js/menus.js]: Other function calls in this file don't add spaces before and after the list of parameters.

File: assets/dev/chat-embedded.html [hunk 4]
```diff
@@ -0,0 +1,37 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+    <!-- Used for testing the effects of embedding the chat -->
+    <!-- Partically for the Notification API -->
+	<meta charset="UTF-8">
+	<meta name="viewport" content="width=device-width, initial-scale=1.0">
+	<meta http-equiv="X-UA-Compatible" content="ie=edge">
+	<title>Embeded chat - FOR TESTING ONLY</title>
```
Comment (KenanY) [inline assets/dev/chat-embedded.html]: "Embeded" => "Embedded"

File: assets/dev/chat-embedded.html [hunk 5]
```diff
@@ -0,0 +1,37 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+    <!-- Used for testing the effects of embedding the chat -->
+    <!-- Partically for the Notification API -->
```
Comment (KenanY) [inline assets/dev/chat-embedded.html]: "Partically" => "Particularly"

File: assets/dev/chat-embedded.html [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+	<meta charset="UTF-8">
+	<meta name="viewport" content="width=device-width, initial-scale=1.0">
+	<meta http-equiv="X-UA-Compatible" content="ie=edge">
+	<title>Embeded chat - FOR TESTING ONLY</title>
+	<style>
+		html {
```
Comment (KenanY) [inline assets/dev/chat-embedded.html]: 4-space indent?

File: assets/dev/chat-embedded.html [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+		.main	{
+			width: 70%;
+		}
+
+		.side	{
+            flex-grow: 1;
+            min-width: 400px;
+		}
+
```
Comment (KenanY) [inline assets/dev/chat-embedded.html]: Remove blank line

File: assets/notification-request/notification-request.html [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+	<title>Notification Permission Request - Strims</title>
+	<meta charset="utf-8">
+	<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1" charset="utf-8">
+	<meta name="referrer" content="no-referrer">
+	<link rel="icon" href="/img/favicon.png">
+	<link rel="stylesheet" href="/notification-request.css">
+	<script src="/notification-request.js"></script>
+</head>
+<body>
+
```
Comment (KenanY) [inline assets/notification-request/notification-request.html]: Remove blank line

File: assets/notification-request/notification-request.html [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    <div class="persona-guide" style="visibility: hidden">
+        <div class="text-bubble">
+            <div class="text-bubble-block">
+                <span class="text-bubble-text"></span>
+                <div class="text-bubble-cursor"></div>
+            </div>
+            <div class="text-bubble-tail"></div>
+        </div>
+        <img src="/img/persona.png" alt="" class="persona">
```
Comment (KenanY) [inline assets/notification-request/notification-request.html]: Finish `alt` tag

File: assets/notification-request/script.js [hunk 10]
```diff
@@ -0,0 +1,164 @@
+const TEXT_TYPING_TIME = 50;
+const TEXT_CLEAR_TIME = 35;
+const CLOSE_BUFFER_TIME = 1700;
+
+const DIALOG = {
+	default: "Requesting permission...",
```
Comment (KenanY) [inline assets/notification-request/script.js]: 4-space indent?

File: assets/notification-request/script.js [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+window.addEventListener( "load", () => {
+    if( Notification.permission === "denied" ) {
+        alreadySetScenario();
+    } else {
+        defaultScenario();
+    }
+} );
+
+/**
+ * Encompasses logic for handeling the sernario where
```
Comment (KenanY) [inline assets/notification-request/script.js]: "handeling" => "handling"

Comment (KenanY) [inline assets/notification-request/script.js]: "sernario" => "scenario"

File: assets/notification-request/script.js [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+ * the notification permission is default/ask.
+ */
+function defaultScenario() {
+    const guideEl = document.querySelector(".persona-guide");
+    guideEl.style = ""; // Overwrite `visibility: hidden`
+
+	typeText( DIALOG.default );
```
Comment (KenanY) [inline assets/notification-request/script.js]: This line's indent does not match that of the lines above it.

File: assets/notification-request/script.js [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+			let response = "";
+
+			switch( result ) {
+				case "granted":
+					const personaEl = document.querySelector(".persona");
+					personaEl.style = "animation-name: persona-excite";
+					response = DIALOG.confirm;
+					break;
+				case "default":
+                case "denied":
```
Comment (KenanY) [inline assets/notification-request/script.js]: Wrong indent?

File: assets/notification-request/script.js [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+
+/**
+ * Fills persona text one letter at a time.
+ */
+function typeText(text) {
+    return new Promise( (resolve, reject) => {
+        const textEl = document.querySelector(".text-bubble-text");
+        const cursorEl = document.querySelector(".text-bubble-cursor");
+
+		let count = 1;
```
Comment (KenanY) [inline assets/notification-request/script.js]: Wrong indent?

File: assets/notification-request/style.scss [hunk 15]
```diff
@@ -0,0 +1,101 @@
+html {
+	background: #222;
+}
+
+body {
+    margin: 0;
+	font-family: "Roboto", Helvetica, "Trebuchet MS", Verdana, sans-serif;
```
Comment (KenanY) [inline assets/notification-request/style.scss]: Mismatched indent

File: assets/notification-request/style.scss [hunk 16]
```diff
@@ -0,0 +1,101 @@
+html {
+	background: #222;
```
Comment (KenanY) [inline assets/notification-request/style.scss]: 4-space indent like the rest of this file?

File: assets/tld.json [hunk 17]
```diff
@@ -1 +0,0 @@
..."AM","AO","AQ","AR","AS","AT","AU","AW","AX","AZ","BA","BB","BD","BE","BF","BG","BH","BI","BJ","BM","BN","BO","BR","BS","BT","BV","BW","BY","BZ","CA","CC","CD","CF","CG","CH","CI","CK","CL","CM","CN","CO","CR","CU","CV","CW","CX","CY","CZ","DE","DJ","DK","DM","DO","DZ","EC","EE","EG","ER","ES","ET","EU","FI","FJ","FK","FM","FO","FR","GA","GB","GD","GE","GF","GG","GH","GI","GL","GM","GN","GP","GQ","GR","GS","GT","GU","GW","GY","HK","HM","HN","HR","HT","HU","ID","IE","IL","IM","IN","IO","IQ","IR","IS","IT","JE","JM","JO","JP","KE","KG","KH","KI","KM","KN","KP","KR","KW","KY","KZ","LA","LB","LC","LI","LK","LR","LS","LT","LU","LV","LY","MA","MC","MD","ME","MG","MH","MK","ML","MM","MN","MO","MP","MQ","MR","MS","MT","MU","MV","MW","MX","MY","MZ","NA","NC","NE","NF","NG","NI","NL","NO","NP","NR","NU","NZ","OM","PA","PE","PF","PG","PH","PK","PL","PM","PN","PR","PS","PT","PW","PY","QA","RE","RO","RS","RU","RW","SA","SB","SC","SD","SE","SG","SH","SI","SJ","SK","SL","SM","SN","SO","SR","ST","SU","SV","SX","SY","SZ","TC","TD","TF","TG","TH","TJ","TK","TL","TM","TN","TO","TR","TT","TV","TW","TZ","UA","UG","UK","US","UY","UZ","VA","VC","VE","VG","VI","VN","VU","WF","WS","YE","YT","ZA","ZM","ZW"]
```
Comment (KenanY) [inline assets/tld.json]: Should probably not be deleted?

File: assets/chat/js/menus.js (hunk 1 above)
Comment (JoannesJ) [reply assets/chat/js/menus.js]: I see. 
I think the next best thing would be a user-agent library with a whitelist.
So the custom popup would be the default action, and browsers on the whitelist would get a direct permission request.

File: assets/tld.json (hunk 17 above)
Comment (JoannesJ) [reply assets/tld.json]: This file is generated on build (scripts/tld.js) and was already included in the gitignore file.
So I figured it was there by accident and would only generate noise.

**Comment by JoannesJ** [comment]
All requested changes resolved.
The last commit is new.
Ready for review.

**Comment by KenanY** [comment]
Some screenshots so others can see what it looks like

image
image

---

**Comment by JoannesJ** [comment]
Hmmm, I hadn't considered private browsing in chrome.
Turns out they set permission to `default` and return `denied` on request without actually asking the user.
Detecting private browsing seems a bit sketchy, and considering this scenario is rare, I would leave it as is.

Am I dumb, or is this absolute garbage API?

Btw, what tool do you use for a video screen capture.

**Comment by KenanY** [comment]
Nah, just leave incognito mode as-is :+1: Tool is ShareX

**Comment by xDashh** [comment]
Merged in https://github.com/MemeLabs/chat-gui/commit/b6ffb66ee4f892012f5d35fd36cb62c2934e88df thanks a lot!

**Pull request closed without merging**